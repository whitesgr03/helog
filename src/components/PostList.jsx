import style from "../styles/PostList.module.css";

import image from "../styles/utils/image.module.css";

import PropTypes from "prop-types";
import { format } from "date-fns";

import { useOutletContext, Link } from "react-router-dom";

const Post = ({ post, darkTheme }) => (
	<>
		<Link to={`/posts/${post.id}`}>
			<div className={`${darkTheme ? image.dark : ""} ${image.content}`}>
				<img src={post.url} alt={post.title} />
			</div>
		</Link>
		<div className={style.container}>
			<strong className={style.dateTime}>
				{format(post.createdAt, "MMMM d, y")}
			</strong>
			<Link to={`/posts/${post.id}`}>
				<h3 className={style.title}>{post.title}</h3>
			</Link>
			<p className={style.content}>{post.content}</p>
		</div>
	</>
);

const PostList = () => {
	const darkTheme = true;
	const { posts } = useOutletContext();

	const items = posts.map(post => (
		<li key={post.id} className={style.item}>
			<Post post={post} />
		</li>
	));

	return (
		<ul className={`${darkTheme ? style.dark : ""}  ${style.postList}`}>
			{items}
		</ul>
	);
};

Post.propTypes = {
	post: PropTypes.object,
	darkTheme: PropTypes.bool,
};

PostList.propTypes = {
	posts: PropTypes.array,
};

export { PostList as default, Post };
