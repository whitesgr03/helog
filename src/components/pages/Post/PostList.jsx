// Modules
import { useRef, useEffect } from 'react';
import { useOutletContext } from 'react-router-dom';
import { useInfiniteQuery } from '@tanstack/react-query';

// Styles
import styles from './PostList.module.css';
import buttonStyles from '../../../styles/button.module.css';

// Components
import { Posts } from './Posts';
import { Loading } from '../../utils/Loading';

// Utils
import { infiniteQueryPostsOption } from '../../../utils/queryOptions';

export const PostList = () => {
	const { onAlert } = useOutletContext();

	const {
		isPending,
		isError,
		data,
		refetch,
		fetchNextPage,
		isFetchingNextPage,
		hasNextPage,
	} = useInfiniteQuery({
		...infiniteQueryPostsOption,
		retry: (failureCount, error) => {
			failureCount >= 3 &&
				error &&
				onAlert({
					message:
						'Loading the posts has some errors occur, please try again later.',
					error: false,
					delay: 4000,
				});
			return failureCount < 3;
		},
	});

	const postListRef = useRef(null);

	const posts = data?.pages.reduce(
		(accumulator, current) => accumulator.concat(current.data.posts),
		[],
	);

	useEffect(() => {
		const handleScroll = async () => {
			const targetRect = postListRef.current.getBoundingClientRect();

			const isScrollToBottom = targetRect.bottom <= window.innerHeight;

			!isFetchingNextPage && isScrollToBottom && fetchNextPage();
		};

		!isError && hasNextPage && window.addEventListener('scroll', handleScroll);
		return () => window.removeEventListener('scroll', handleScroll);
	}, [isError, hasNextPage, isFetchingNextPage, fetchNextPage]);

	return (
		<div className={styles['post-list']}>
			{isError && !data?.pages.length ? (
				<button
					className={`${buttonStyles.content} ${buttonStyles.more}`}
					onClick={() => refetch()}
				>
					Click here to load posts
				</button>
			) : isPending ? (
				<Loading text="Loading Posts..." />
			) : (
				<>
					<div className={styles.container} ref={postListRef}>
						<Posts posts={posts} />
					</div>
					{isFetchingNextPage && <Loading text={'Loading more posts ...'} />}
				</>
			)}
		</div>
	);
};
