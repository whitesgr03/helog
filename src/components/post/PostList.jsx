// Styles
import style from '../../styles/post/PostList.module.css';

// Components
import { Posts } from './Posts';

export const PostList = () => {
	return (
		<div className={style.postList}>
			<Posts />
		</div>
	);
};
