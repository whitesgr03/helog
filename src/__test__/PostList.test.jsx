import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { BrowserRouter, Outlet, Route, Routes } from "react-router-dom";
import { format } from "date-fns";

import PostList from "../components/PostList";

describe("PostList component", () => {
	it("should render correct data length", () => {
		const mockPosts = [
			{
				id: "1",
				url: "#",
				title: "This is title A",
				content: "This is content A",
				createdAt: new Date("2024/5/1"),
			},
			{
				id: "2",
				url: "#",
				title: "This is title B",
				content: "This is content B",
				createdAt: new Date("2024/5/1"),
			},
			{
				id: "3",
				url: "#",
				title: "This is title C",
				content: "This is content C",
				createdAt: new Date("2024/5/1"),
			},
		];

		render(
			<BrowserRouter>
				<Routes>
					<Route element={<Outlet context={{ posts: mockPosts }} />}>
						<Route index element={<PostList />} />
					</Route>
				</Routes>
			</BrowserRouter>
		);

		const actual = screen.getAllByRole("listitem");

		expect(actual.length).equal(3);
	});
	it("should render image alt and url from props", () => {
		const mockPosts = [
			{
				id: "1",
				url: "#",
				title: "This is title",
				content: "This is content",
				createdAt: new Date("2024/5/1"),
			},
		];

		render(
			<BrowserRouter>
				<Routes>
					<Route element={<Outlet context={{ posts: mockPosts }} />}>
						<Route index element={<PostList />} />
					</Route>
				</Routes>
			</BrowserRouter>
		);

		const actual = screen.getByAltText(mockPosts[0].title);

		expect(actual).toHaveAttribute("src", mockPosts.url);
	});
	it("should render createdAt from props", () => {
		const mockPosts = [
			{
				id: "1",
				url: "#",
				title: "This is title",
				content: "This is content",
				createdAt: new Date("2024/5/1"),
			},
		];
		render(
			<BrowserRouter>
				<Routes>
					<Route element={<Outlet context={{ posts: mockPosts }} />}>
						<Route index element={<PostList />} />
					</Route>
				</Routes>
			</BrowserRouter>
		);

		const actual = screen.queryByText(
			format(mockPosts[0].createdAt, "MMMM d, y")
		);

		expect(actual).not.equal(null);
	});
	it("should render title from props", () => {
		const mockPosts = [
			{
				id: "1",
				url: "#",
				title: "This is title",
				content: "This is content",
				createdAt: new Date("2024/5/1"),
			},
		];
		render(
			<BrowserRouter>
				<Routes>
					<Route element={<Outlet context={{ posts: mockPosts }} />}>
						<Route index element={<PostList />} />
					</Route>
				</Routes>
			</BrowserRouter>
		);

		const actual = screen.queryByRole("heading", {
			name: mockPosts[0].title,
		});

		expect(actual).not.equal(null);
	});
	it("should render content from props", () => {
		const mockPosts = [
			{
				id: "1",
				url: "#",
				title: "This is title",
				content: "This is content",
				createdAt: new Date("2024/5/1"),
			},
		];
		render(
			<BrowserRouter>
				<Routes>
					<Route element={<Outlet context={{ posts: mockPosts }} />}>
						<Route index element={<PostList />} />
					</Route>
				</Routes>
			</BrowserRouter>
		);

		const actual = screen.queryByText(mockPosts[0].content);

		expect(actual).not.equal(null);
	});
});
