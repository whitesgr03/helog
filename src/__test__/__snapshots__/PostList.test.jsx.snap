// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostList component > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="_postList_05bdc4"
  />
</DocumentFragment>
`;
